examples/smart_contract.ts

```diff
@@ -0,0 +1,277 @@
+      callData,
+    )
+      .next((encodedResponse) => {
+        const response = $contractsApiCallResult.decode(U.hex.decode(encodedResponse))
+        if (message.returnType === null) {
+          return undefined
+        }
+        return $result.decode(response.result.data)
+      })
+  }
+
+  tx<Args extends unknown[]>(
+    origin: Uint8Array,
+    messageLabel: string,
+    args: Args,
+    sign: C.Z.$<C.M.Signer>,
+  ) {
+    const message = this.#getMessageByLabel(messageLabel)!
+    const [$args, _] = this.#getMessageCodecs(message)
+    const data = $args.encode([U.hex.decode(message.selector), ...args])
+    const value = this.#preSubmitContractCallDryRunGasEstimate(origin, data)
+      .next(({ gasRequired }) => {
+        return {
+          type: "call",
+          dest: C.MultiAddress.Id(this.contractAddress),
+          value: 0n,
+          data,
+          gasLimit: {
+            refTime: gasRequired.refTime,
+            proofSize: gasRequired.proofSize,
+          },
+          storageDepositLimit: undefined,
+        }
+      })
+    const tx = C.extrinsic(client)({
+      sender: C.MultiAddress.Id(origin),
+      call: C.Z.rec({
+        type: "Contracts",
+        value,
+      }),
+    })
+      .signed(sign)
+    const finalizedIn = tx.watch(({ end }) =>
+      (status) => {
+        if (typeof status !== "string" && (status.inBlock ?? status.finalized)) {
+          return end(status.inBlock ?? status.finalized)
+        } else if (C.rpc.known.TransactionStatus.isTerminal(status)) {
+          return end(new Error())
+        }
+        return
+      }
+    )
+    return C.Z.ls(finalizedIn, C.events(tx, finalizedIn))
+  }
+
+  // TODO: codegen each contract message as a method
+
+  #getMessageByLabel(label: string) {
+    return this.metadata.V3.spec.messages.find((c) => c.label === label)
+  }
+
+  #getMessageCodecs(
+    message: C.M.ContractMetadata.Message,
+  ): [C.$.Codec<unknown[]>, C.$.Codec<any>] {
+    const argCodecs = [
+      // message selector
+      C.$.sizedUint8Array(U.hex.decode(message.selector).length),
+      // message args
+      ...message.args.map((arg) => this.deriveCodec(arg.type.type)),
+    ]
+    const $result = message.returnType !== null
+      ? this.deriveCodec(message.returnType.type)
+      : C.$.constant(null)
+    // @ts-ignore ...
+    return [C.$.tuple(...argCodecs), $result]
+  }
+
+  #preSubmitContractCallDryRunGasEstimate(
+    origin: Uint8Array,
+    data: Uint8Array,
+  ) {
+    const callData = U.hex.encode($contractsApiCallArgs.encode([
+      origin,
+      this.contractAddress,
+      0n,
+      undefined,
+      undefined,
+      data,
+    ]))
+    return C.state.call(client)(
+      "ContractsApi_call",
+      callData,
+    )
+      .next((encodedResponse) => {
+        return $contractsApiCallResult.decode(U.hex.decode(encodedResponse))
+      })
+  }
+}
+
+const flipperContract = new SmartContract(contract.metadata, contractAddress)
```

Comment:
The following contract calls could be simplified without `console.logs` like
```
await flipperContract.call(T.alice.publicKey, "get", []).run()
await flipperContract.tx(T.alice.publicKey, "flip", [], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get", []).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.tx(T.alice.publicKey, "inc", [], T.alice.sign).run()
await flipperContract.tx(T.alice.publicKey, "inc", [], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.tx(T.alice.publicKey, "inc_by", [3], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.call(T.alice.publicKey, "m1", [2, true]).run()
await flipperContract.call(T.alice.publicKey, "m1", [3, false]).run()
```

For messages that mutate the state, `contract.tx(origin, message, args, sign)` is used.
For other messages, `contract.call(origin, message, args)` is used.

The source code for this contract is
```
#[ink::contract]
mod flipper {

    /// Defines the storage of your contract.
    /// Add new fields to the below struct in order
    /// to add new static storage fields to your contract.
    #[ink(storage)]
    pub struct Flipper {
        /// Stores a single `bool` value on the storage.
        value: bool,
        /// Stores count
        count: i32,
    }

    #[derive(Debug, scale::Encode, scale::Decode)]
    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
    pub struct M2 {
        a: u32,
        b: bool,
    }

    impl Flipper {
        /// Constructor that initializes the `bool` value to the given `init_value`.
        #[ink(constructor)]
        pub fn new(init_value: bool) -> Self {
            Self {
                value: init_value,
                count: 0,
            }
        }

        /// Constructor that initializes the `bool` value to `false`.
        ///
        /// Constructors can delegate to other constructors.
        #[ink(constructor)]
        pub fn default() -> Self {
            Self::new(Default::default())
        }

        /// A message that can be called on instantiated contracts.
        /// This one flips the value of the stored `bool` from `true`
        /// to `false` and vice versa.
        #[ink(message)]
        pub fn flip(&mut self) {
            self.value = !self.value;
        }

        /// Simply returns the current value of our `bool`.
        #[ink(message)]
        pub fn get(&self) -> bool {
            self.value
        }

        /// multiple arg method1.
        #[ink(message)]
        pub fn m1(&self, a: u32, b: bool) -> (u32, bool) {
            (a, b)
        }

        /// multiple arg method2.
        #[ink(message)]
        pub fn m2(&self, a: u32, b: bool) -> M2 {
            M2 { a, b }
        }

        /// get the current count.
        #[ink(message)]
        pub fn get_count(&self) -> i32 {
            self.count
        }

        /// increment current count by 1
        #[ink(message)]
        pub fn inc(&mut self) {
            self.inc_by(1);
        }

        /// decrement current count by 1
        #[ink(message)]
        pub fn dec(&mut self) {
            self.inc_by(-1);
        }

        /// increment current count by n
        #[ink(message)]
        pub fn inc_by(&mut self, n: i32) {
            self.count += n;
        }
    }
```